src/diffusers/pipelines/modular_pipeline.py

```diff
@@ -1352,340 +1280,257 @@ def dtype(self) -> torch.dtype:
-    #         output += f"Trigger Inputs: {block.trigger_inputs}\n"
-    #         # Get first trigger input as example
-    #         example_input = next(t for t in block.trigger_inputs if t is not None)
-    #         output += f"  Use `get_execution_blocks()` with input names to see selected blocks (e.g. `get_execution_blocks('{example_input}')`).\n"
-    #         output += "Check `.doc` of returned object for more information.\n\n"
-
-    #     # List the call parameters
-    #     full_doc = self.pipeline_block.doc
-    #     if "------------------------" in full_doc:
-    #         full_doc = full_doc.split("------------------------")[0].rstrip()
-    #     output += full_doc
-
-    #     return output
-
-    # YiYi TODO: try to unify the to method with the one in DiffusionPipeline
-    # Modified from diffusers.pipelines.pipeline_utils.DiffusionPipeline.to
-    def to(self, *args, **kwargs):
-        r"""
-        Performs Pipeline dtype and/or device conversion. A torch.dtype and torch.device are inferred from the
-        arguments of `self.to(*args, **kwargs).`
-
-        <Tip>
-
-            If the pipeline already has the correct torch.dtype and torch.device, then it is returned as is. Otherwise,
-            the returned pipeline is a copy of self with the desired torch.dtype and torch.device.
-
-        </Tip>
+        components_to_register = {}
+        for name in components_to_load:
+            spec = self._component_specs[name]
+            component_load_kwargs = {}
+            for key, value in kwargs.items():
+                if not isinstance(value, dict):
+                    # if the value is a single value, apply it to all components
+                    component_load_kwargs[key] = value
+                else:
+                    if name in value:
+                        # if it is a dict, check if the component name is in the dict
+                        component_load_kwargs[key] = value[name]
+                    elif "default" in value:
+                        # check if the default is specified
+                        component_load_kwargs[key] = value["default"]
+            try:
+                components_to_register[name] = spec.create(**component_load_kwargs)
+            except Exception as e:
+                logger.warning(f"Failed to create component '{name}': {e}")
+            
+        # Register all components at once
+        self.register_components(**components_to_register)
 
+    # YiYi TODO: should support to method
+    def to(self, *args, **kwargs):
+        pass
+
+    # YiYi TODO: 
+    # 1. should support save some components too! currently only modular_model_index.json is saved
+    # 2. maybe order the json file to make it more readable: configs first, then components
+    def save_pretrained(self, save_directory: Union[str, os.PathLike], push_to_hub: bool = False, spec_only: bool = True, **kwargs):
```

Comment:
TODO:
currently I only save config (a.k.a the model_index), we could totally save selected components into an subfolder too, so for example fine tune trainers could make a modular repo that contains fine tuned components and with the modular_model_index points to other repos for frozen components

same goes for from_pretrained()